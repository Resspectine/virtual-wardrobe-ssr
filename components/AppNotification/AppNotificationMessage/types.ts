import { Notification } from '@/types/notification';

export interface AppNotificationMessageProps {
  notification: Notification;
  onRemove: (id: number) => void;
}
