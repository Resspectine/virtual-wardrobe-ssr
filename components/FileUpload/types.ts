import { FieldErrors, UseFormRegister, UseFormSetValue } from 'react-hook-form';

import { CreateGarmentValues } from '@/pageComponents/CreateGarment/types';

export interface FileUploadProps {
  register: UseFormRegister<CreateGarmentValues>;
  name: keyof CreateGarmentValues;
  file: File;
  setValue: UseFormSetValue<CreateGarmentValues>;
  disabled?: boolean;
  errors: FieldErrors<CreateGarmentValues>;
}
