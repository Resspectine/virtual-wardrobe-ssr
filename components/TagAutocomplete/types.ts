import { AutocompleteProps } from '@mui/material';

import { Tag } from 'types/tag';

export type CustomTag = { inputValue: string } & Omit<Tag, 'id'>;

export type AnyTag = Tag | CustomTag;

export type TagAutocompleteProps = AutocompleteProps<AnyTag, true, undefined, false>;
