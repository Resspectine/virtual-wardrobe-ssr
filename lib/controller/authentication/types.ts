import { User } from '@/types/user';

export type LoginUser = Pick<User, 'email' | 'password'>;

export type RegisterUser = Pick<User, 'email' | 'password' | 'name'>;
