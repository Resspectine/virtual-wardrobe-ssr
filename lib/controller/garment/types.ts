import { Garment } from '@/types/garment';
import { Tag } from '@/types/tag';

export interface LoadGarmentsQueryParams {
  filter?: string;
  orderBy?: string;
  orderDirection?: string;
}

export type CreateRequestGarment = Omit<Garment, 'id' | 'tags'> & { tags: Omit<Tag, 'id'>[] };

export type UpdateRequestGarment = Omit<Garment, 'tags'> & { tags: (Omit<Tag, 'id'> & { id?: string })[] };
