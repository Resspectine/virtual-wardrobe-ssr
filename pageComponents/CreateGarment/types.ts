export interface CreateGarmentValues {
  title: string;
  description: string;
  price: string;
  image: File[];
}
