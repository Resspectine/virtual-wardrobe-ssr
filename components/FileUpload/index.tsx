import Close from '@mui/icons-material/Close';
import { FormHelperText } from '@mui/material';
import Box from '@mui/material/Box';
import Button from '@mui/material/Button';
import { FC } from 'react';

import { useFileUploadControl } from './hooks';
import { CloseIcon, Image } from './styled';
import { FileUploadProps } from './types';

const FileUpload: FC<FileUploadProps> = ({ register, name, file, setValue, disabled, errors }) => {
  const { fileUrl, onCloseCLick, error } = useFileUploadControl({
    file,
    name,
    setValue,
    errors,
  });

  return fileUrl ? (
    <Box position="relative">
      <Image component="img" src={fileUrl} alt="Something went wrong" />
      <CloseIcon onClick={onCloseCLick}>
        <Close />
      </CloseIcon>
    </Box>
  ) : (
    <>
      <Button variant="outlined" component="label" disabled={disabled}>
        Upload File
        <input
          type="file"
          hidden
          {...register(name, {
            required: 'Image is required',
          })}
        />
      </Button>
      {error && (
        <FormHelperText error>
          {Array.isArray(error) ? error.map(errorMessage => errorMessage?.message).join(', ') : error.message}
        </FormHelperText>
      )}
    </>
  );
};

export default FileUpload;
