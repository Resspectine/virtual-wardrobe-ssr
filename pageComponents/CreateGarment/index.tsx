import Box from '@mui/material/Box';
import Typography from '@mui/material/Typography';
import { FC } from 'react';

import { getFormFieldConfigurations } from './helpers';
import { useCreateGarment } from './hooks';
import { CreateGarmentForm, CreateGarmentSubmit, CreateGarmentTextField } from './styled';

import FileUpload from '@/components/FileUpload';
import Autocomplete from '@/components/TagAutocomplete';

const CreateGarment: FC = () => {
  const {
    tags,
    watch,
    errors,
    control,
    isValid,
    onSubmit,
    register,
    setValue,
    autocompleteValue,
    setAutocompleteValue,
  } = useCreateGarment();

  return (
    <Box>
      <Typography>Create garment</Typography>
      <CreateGarmentForm component="form" onSubmit={onSubmit}>
        {getFormFieldConfigurations(control).map((props, index) => (
          <CreateGarmentTextField {...props} key={index} />
        ))}
        {tags && <Autocomplete autocompleteOptions={tags} setValue={setAutocompleteValue} value={autocompleteValue} />}
        <FileUpload register={register} name="image" file={watch('image')[0]} setValue={setValue} errors={errors} />
        <CreateGarmentSubmit type="submit" variant="contained" color="primary" disabled={!isValid}>
          Submit
        </CreateGarmentSubmit>
      </CreateGarmentForm>
    </Box>
  );
};

export default CreateGarment;
