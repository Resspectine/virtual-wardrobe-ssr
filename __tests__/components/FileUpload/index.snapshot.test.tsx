import { act, waitFor } from '@testing-library/react';
import { renderHook } from '@testing-library/react-hooks';
import userEvent from '@testing-library/user-event';
import { useForm } from 'react-hook-form';

import FileUpload from '@/components/FileUpload';
import { CreateGarmentValues } from '@/pageComponents/CreateGarment/types';
import { render } from '@/testUtils';

describe('FileUpload', () => {
  test('should show upload button', async () => {
    const { result } = renderHook(() =>
      useForm<CreateGarmentValues>({
        defaultValues: {
          image: [],
          title: '',
          description: '',
          price: '',
        },
        mode: 'onChange',
      })
    );

    const { asFragment } = render(
      <FileUpload
        errors={result.current.formState.errors}
        file={result.current.watch('image')[0]}
        name="image"
        register={result.current.register}
        setValue={result.current.setValue}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test('should show picture', async () => {
    const file = new File([], 'test');

    const { result } = renderHook(() =>
      useForm<CreateGarmentValues>({
        defaultValues: {
          image: [],
          title: '',
          description: '',
          price: '',
        },
        mode: 'onChange',
      })
    );

    const { asFragment, getByRole, rerender, findByAltText } = render(
      <FileUpload
        errors={result.current.formState.errors}
        file={result.current.watch('image')[0]}
        name="image"
        register={result.current.register}
        setValue={result.current.setValue}
      />
    );

    userEvent.upload(getByRole('button'), file);

    rerender(
      <FileUpload
        errors={result.current.formState.errors}
        file={result.current.watch('image')[0]}
        name="image"
        register={result.current.register}
        setValue={result.current.setValue}
      />
    );

    await waitFor(async () => expect(await findByAltText('Something went wrong')).toBeInTheDocument());

    expect(asFragment()).toMatchSnapshot();
  });

  test('should show error', async () => {
    const { result } = renderHook(() =>
      useForm<CreateGarmentValues>({
        defaultValues: {
          image: [],
          title: '',
          description: '',
          price: '',
        },
        mode: 'onChange',
      })
    );

    const { asFragment, rerender } = render(
      <FileUpload
        errors={result.current.formState.errors}
        file={result.current.watch('image')[0]}
        name="image"
        register={result.current.register}
        setValue={result.current.setValue}
      />
    );

    await act(async () => {
      await result.current.handleSubmit(() => {})();
    });

    rerender(
      <FileUpload
        errors={result.current.formState.errors}
        file={result.current.watch('image')[0]}
        name="image"
        register={result.current.register}
        setValue={result.current.setValue}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
