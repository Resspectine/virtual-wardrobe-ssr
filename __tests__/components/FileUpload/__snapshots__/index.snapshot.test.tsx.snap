// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileUpload should show error 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  position: relative;
  box-sizing: border-box;
  -webkit-tap-highlight-color: transparent;
  background-color: transparent;
  outline: 0;
  border: 0;
  margin: 0;
  border-radius: 0;
  padding: 0;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  -moz-appearance: none;
  -webkit-appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
  font-family: Montserrat,Times New Roman;
  font-weight: 500;
  font-size: 0.875rem;
  line-height: 1.75;
  text-transform: uppercase;
  min-width: 64px;
  padding: 5px 15px;
  border-radius: 4px;
  -webkit-transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,border-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,border-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  border: 1px solid rgba(25, 118, 210, 0.5);
  color: #1976d2;
}

.emotion-0::-moz-focus-inner {
  border-style: none;
}

.emotion-0.Mui-disabled {
  pointer-events: none;
  cursor: default;
}

@media print {
  .emotion-0 {
    -webkit-print-color-adjust: exact;
    color-adjust: exact;
  }
}

.emotion-0:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: rgba(25, 118, 210, 0.04);
  border: 1px solid #1976d2;
}

@media (hover: none) {
  .emotion-0:hover {
    background-color: transparent;
  }
}

.emotion-0.Mui-disabled {
  color: rgba(0, 0, 0, 0.26);
  border: 1px solid rgba(0, 0, 0, 0.12);
}

.emotion-1 {
  overflow: hidden;
  pointer-events: none;
  position: absolute;
  z-index: 0;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  border-radius: inherit;
}

<label
    class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root  emotion-0"
    role="button"
    tabindex="0"
  >
    Upload File
    <input
      hidden=""
      name="image"
      type="file"
    />
    <span
      class="MuiTouchRipple-root emotion-1"
    />
  </label>
  .emotion-0 {
  color: #388e3c;
  font-family: Montserrat,Times New Roman;
  font-weight: 400;
  font-size: 0.75rem;
  line-height: 1.66;
  text-align: left;
  margin-top: 3px;
  margin-right: 0;
  margin-bottom: 0;
  margin-left: 0;
}

.emotion-0.Mui-disabled {
  color: rgba(0, 0, 0, 0.38);
}

.emotion-0.Mui-error {
  color: #d32f2f;
}

<p
    class="MuiFormHelperText-root Mui-error emotion-0"
  >
    Image is required
  </p>
</DocumentFragment>
`;

exports[`FileUpload should show picture 1`] = `
<DocumentFragment>
  .emotion-0 {
  position: relative;
}

.emotion-1 {
  max-width: 300px;
  height: auto;
}

.emotion-2 {
  position: absolute;
  top: 5px;
  right: 5px;
  background-color: #fff;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border-radius: 8px;
}

.emotion-3 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1em;
  height: 1em;
  display: inline-block;
  fill: currentColor;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  font-size: 1.5rem;
}

<div
    class="MuiBox-root emotion-0"
  >
    <img
      alt="Something went wrong"
      class="MuiBox-root emotion-1"
      src="data:application/octet-stream;base64,"
    />
    <div
      class="MuiBox-root emotion-2"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium emotion-3"
        data-testid="CloseIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;

exports[`FileUpload should show upload button 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  position: relative;
  box-sizing: border-box;
  -webkit-tap-highlight-color: transparent;
  background-color: transparent;
  outline: 0;
  border: 0;
  margin: 0;
  border-radius: 0;
  padding: 0;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  -moz-appearance: none;
  -webkit-appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
  font-family: Montserrat,Times New Roman;
  font-weight: 500;
  font-size: 0.875rem;
  line-height: 1.75;
  text-transform: uppercase;
  min-width: 64px;
  padding: 5px 15px;
  border-radius: 4px;
  -webkit-transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,border-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,border-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  border: 1px solid rgba(25, 118, 210, 0.5);
  color: #1976d2;
}

.emotion-0::-moz-focus-inner {
  border-style: none;
}

.emotion-0.Mui-disabled {
  pointer-events: none;
  cursor: default;
}

@media print {
  .emotion-0 {
    -webkit-print-color-adjust: exact;
    color-adjust: exact;
  }
}

.emotion-0:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: rgba(25, 118, 210, 0.04);
  border: 1px solid #1976d2;
}

@media (hover: none) {
  .emotion-0:hover {
    background-color: transparent;
  }
}

.emotion-0.Mui-disabled {
  color: rgba(0, 0, 0, 0.26);
  border: 1px solid rgba(0, 0, 0, 0.12);
}

.emotion-1 {
  overflow: hidden;
  pointer-events: none;
  position: absolute;
  z-index: 0;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  border-radius: inherit;
}

<label
    class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root  emotion-0"
    role="button"
    tabindex="0"
  >
    Upload File
    <input
      hidden=""
      name="image"
      type="file"
    />
    <span
      class="MuiTouchRipple-root emotion-1"
    />
  </label>
</DocumentFragment>
`;
